src/Microsoft.Management.Deployment/PackageCatalogProgress.cpp

```diff
@@ -0,0 +1,137 @@
+                    progressWeights.insert_or_assign(AppInstaller::ProgressType::Bytes, 0.7);
+                    progressWeights.insert_or_assign(AppInstaller::ProgressType::Percent, 0.3);
+                }
+
+                return std::make_shared<PreIndexedPackageCatalogProgressSink>(progressWeights, progressReporter);
+            }
+            else
+            {
+                return std::make_shared<CompletionOnlyProgressSink>(progressReporter);
+            }
+        }
+    }
+
+    CompletionOnlyProgressSink::CompletionOnlyProgressSink(std::function<void(double)> progressReporter) :
+        m_progressReporter(progressReporter)
+    {
+    }
+
+    void CompletionOnlyProgressSink::OnProgress(uint64_t current, uint64_t maximum, AppInstaller::ProgressType type)
+    {
+        UNREFERENCED_PARAMETER(current);
+        UNREFERENCED_PARAMETER(maximum);
+        UNREFERENCED_PARAMETER(type);
+    }
+
+    void CompletionOnlyProgressSink::SetProgressMessage(std::string_view message)
+    {
+        UNREFERENCED_PARAMETER(message);
+    }
+
+    void CompletionOnlyProgressSink::BeginProgress()
+    {
+        m_progressReporter(0);
+    }
+
+    void CompletionOnlyProgressSink::EndProgress(bool hideProgressWhenDone)
+    {
+        UNREFERENCED_PARAMETER(hideProgressWhenDone);
+        m_progressReporter(100);
+    }
+
+    PreIndexedPackageCatalogProgressSink::PreIndexedPackageCatalogProgressSink(std::unordered_map<AppInstaller::ProgressType, double> progressWeights, std::function<void(double)> progressReporter) :
+        m_progressWeights(progressWeights), m_progressReporter(progressReporter)
+    {
+        // If no weights are provided, default to percent.
+        if (m_progressWeights.empty())
+        {
+            m_progressWeights[AppInstaller::ProgressType::Percent] = 1;
+        }
+
+        // Calculate the total weight.
+        double totalWeight = 0;
+        for (const auto& weight : m_progressWeights)
+        {
+            if (weight.first != AppInstaller::ProgressType::None)
+            {
+                totalWeight += weight.second;
+            }
+        }
+
+        // If the total weight is greater than 1, throw an exception.
+        if (totalWeight != 1.0)
+        {
+            THROW_HR(E_INVALIDARG);
+        }
+    }
+
+    void PreIndexedPackageCatalogProgressSink::OnProgress(uint64_t current, uint64_t maximum, AppInstaller::ProgressType type)
+    {
+        if (maximum == 0 || type == AppInstaller::ProgressType::None)
+        {
+            return;
+        }
+
+        double progress = static_cast<double>(current) / maximum;
+        m_progressValues[type] = progress;
+
+        double totalProgress = 0.0;
+
+        // Calculate the total progress.
+        for (const auto& [progressType, weight] : m_progressWeights)
+        {
+            // Adjust the total progress value based on the weight.
+            totalProgress += m_progressValues[progressType] * weight;

```

Comment:
This algorithm doesn't work right if one of the types never reports any progress (or doesn't report 100% as its last value).  What might be better (but a bit more complicated) is an ordered set of expected types and weights.  Then when calculating the overall value, the entire weight of any previous step is used rather than the reported value.

For this case, it would be similar except if we have seen any percent based progress we would always use the byte based progress weight value:
```
// New type for m_progressWeights to have an order
std::vector<std::pair<type, weight>> m_progressWeights;

doublt totalWeight = 0, calculatedProgress = 0;

for (const auto& [type, weight] : m_progressWeights)
{
    double typeWeight = m_progressValues[type];
    if (typeWeight != 0)
    {
        calculatedProgress = totalWeight;
    }
    totalWeight += weight;
    calculatedProgress += typeWeight * weight;
}
```
If we needed to support multiples of the same type it would be way more complicated.

Reply:
Good catch! I assumed the underlying operation code would always report progress until 100% completion. There's a potential gap when the initial operations complete successfully but stop reporting before full completion. It took a bit of time to understand this edge case, so I added an inline comment to explain why additional computation logic is necessary when applicable.